### ISSUE TITLE
How to restore container after node restarted

### ISSUE THREAD

**Issue by frostynova** [projectcalico/libnetwork-plugin#49]
Docker 1.10.3
Calico: 0.18.0

After a Docker node restarted, i cannot start the containers deployed with calico:

```
Error response from daemon: service endpoint with name flex.worker.worker.v1-i1-d0 already exists
Error: failed to start containers: flex.worker.worker.v1-i1-d0
```

I tried docker network disconnect & connect, not ok.
Also tried rm this container directly, not ok.

So is there any way i can start this container?

**Comment by TrimBiggs** [comment]
Hi @frostynova,  I've got a few questions.
When you say the Docker node is restarted, what is actually restarted: the host machine, the calico/node container, or the deployed container?

Are you using `docker start` to try to start the container?

When you say you tried to `rm` the container, do you mean `docker rm` or `calicoctl container remove` (the `calicoctl container` commands will not work with Docker networks)? Can you remove the container with `docker rm -f flex.worker.worker.v1-i1-d0`?

I'll try to reproduce and get more info once I fully understand what's going on.

**Comment by frostynova** [comment]
@TrimBiggs 

It is docker daemon or host machine restarted

Yes, `docker start` failed and i used `docker rm` to remove the container with no error, but still can not start a new one with the same name.

I browsed the docker etcd storage and found the 'dead' endpoint in the `/docker/`, after remove it, everything is ok. So i guess: If the docker daemon restarted unexpectedly, docker cannot clean the endpoint info, and this will lead the 'dead' container cannot be started again.

**Comment by frostynova** [comment]
Additional information:

The only way i found to restore a 'dead' container is:

* etcdctl rm /docker/network/v1.0/{ network_id }/{ endpoint_id }
* docker rm old container
* docker run new container with the same name and ip

**Comment by frostynova** [comment]
@TrimBiggs Today i sitll dig on this, and update the fix steps in previous comment

**Comment by TrimBiggs** [comment]
@frostynova, thanks for this.  That sounds like a workaround we've seen before.  It's strange that libnetwork has not cleaned this up either.  I will look into this a bit more by the end of this week to try and narrow down the issue.  I want to learn more about whether libnetwork will do this on its own or whether it is strictly when Calico is in use.

**Comment by frostynova** [comment]
@TrimBiggs I have viewed docker network code these days and found:

* If daemon restarted normally, the code show docker daemon will cleanup the network during the containers stopped. 
* But with `systemctl restart docker`, it seems that calico libnetwork plugin exited first, and then docker daemon cannot communicate with it. So the other containers network cannot be cleaned.
* If daemon restarted unexpectedly, there seems no code to clean the network.

**Comment by TrimBiggs** [comment]
Hey @frostynova, just letting you know I'm looking at this today. I've reproduced the issue and am running some manual tests/debugging now.  My suspicion is that either:
 - Docker exits, killing the calico-libnetwork container, and when docker tries to clean up the endpoints it cannot find the plugin to determine how to clean up so it leaves the endpoints as is
 - When Docker restarts it ignores existing endpoints in its datastore and expects the plugin to automatically clean up (possibly a Docker bug or something to raise with Docker at least if this is the case)

I'll let you know how I get on. Thanks for all of your info above.

**Comment by TrimBiggs** [comment]
I just ran the following test:
 - Run calico-node (in docker), etcd (binary), and calico-libnetwork directly on my machine (NOT in docker)
 - Create docker network and docker container with calico ipam
 - Stop docker (libnetwork logs show it releasing the IP)
 - Start docker (no logs in libnetwork)
 - Start container, which starts normally and is given a new IP address

Everything seemed to work as expected in this test. As you suggested, it seems to be an issue with running `calico-libnetwork` as a Docker container since Docker kills the container before requesting cleanup from Calico network plugin, which makes sense since a container should be stopped before removing its endpoint config.

I also found that if I force kill docker, the endpoint is cleaned up when docker restarts.  Again, this worked totally fine when running libnetwork directly on my machine.

**Comment by frostynova** [comment]
@TrimBiggs So seems we need a standalone lob network plugin running method?

And do you test if the whole machine power halted everything is ok?

**Comment by TrimBiggs** [comment]
I've just raised this issue with Docker: https://github.com/docker/docker/issues/22239
Depending on Docker's response, we may be able to continue running the plugin as a container or we'll have to run a standalone network plugin process of some sort.

I did not test with killing the machine power.

**Comment by TrimBiggs** [comment]
I've tested Docker's workaround for this without success.  I've written another couple of comments in the docker/docker issue above.

**Comment by wrouesnel** [comment]
What's the status on this issue? Does running Calico as a separate process (i.e. not a container) resolve it? I've just run straight into it and it's a little bit of a show-stopper if my static IPs will be broken by a sudden node death.

**Comment by TrimBiggs** [comment]
@wrouesnel I've just gotten a response from Docker in docker/docker#22239 that they are working on a fix for this.  I'm intending to continue pushing for the best solution for Calico's libnetwork plugin container.

**Comment by tomdee** [comment]
libnetwork-plugin is now a static go binary, so should be easy to run outside a container.

**Comment by frostynova** [comment]
@tomdee We running dedicated libnetwork-plugin successfully,  but after docker daemon restarted, systemd is blocked.

I trace the system call and found it was blocked by this api call to libnetwork-plugin

`POST /NetworkDriver.CreateEndpoint`

I read the code and found https://github.com/projectcalico/libnetwork-plugin/blob/3538005e755d8d587988a468fa0d2e1e40050d38/driver/network_driver.go#L138 where plugin will call docker api to fetch network info. This is a deadlock since docker daemon has not been started ok.

**Comment by tomdee** [comment]
@frostynova Thanks for reporting this. Generally, this kind of problem is best reported as a new issue though.

**Comment by CMGS** [comment]
docker will store network info in /docker/network/v1.0/network/{networkid}, how about get network name from etcd?
